deploy/ControllerInit.sol

```diff
@@ -0,0 +1,171 @@
+        address              relayer,
+        ControllerInstance   memory controllerInst,
+        AllocatorIlkInstance memory ilkInst,
+        UsdsInstance         memory usdsInst,
+        RateLimitData        memory usdsMintData,
+        RateLimitData        memory usdcToUsdsData,
+        RateLimitData        memory usdcToCctpData,
+        RateLimitData        memory cctpToBaseDomainData
+    )
+        internal
+    {
+        IALMProxy   almProxy   = IALMProxy(controllerInst.almProxy);
+        IRateLimits rateLimits = IRateLimits(controllerInst.rateLimits);
+
+        IMainnetControllerLike controller = IMainnetControllerLike(controllerInst.controller);
+
+        // Step 1: Configure ACL permissions for controller and almProxy
+
+        controller.grantRole(controller.FREEZER(), freezer);
+        controller.grantRole(controller.RELAYER(), relayer);
+
+        almProxy.grantRole(almProxy.CONTROLLER(), address(controller));
+
+        rateLimits.grantRole(rateLimits.CONTROLLER(), address(controller));
+
+        // Step 2: Configure all rate limits for controller, using Base as only domain
+
+        bytes32 domainKeyBase = RateLimitHelpers.makeDomainKey(
+            controller.LIMIT_USDC_TO_DOMAIN(),
+            CCTPForwarder.DOMAIN_ID_CIRCLE_BASE
+        );
+
+        rateLimits.setRateLimitData(controller.LIMIT_USDS_MINT(),    usdsMintData.maxAmount,         usdsMintData.slope);
+        rateLimits.setRateLimitData(controller.LIMIT_USDS_TO_USDC(), usdcToUsdsData.maxAmount,       usdcToUsdsData.slope);
+        rateLimits.setRateLimitData(controller.LIMIT_USDC_TO_CCTP(), usdcToCctpData.maxAmount,       usdcToCctpData.slope);
+        rateLimits.setRateLimitData(domainKeyBase,                   cctpToBaseDomainData.maxAmount, cctpToBaseDomainData.slope);
+
+        // Step 3: Configure almProxy within the allocation system
+
+        IVaultLike(ilkInst.vault).rely(controllerInst.almProxy);
+        IBufferLike(ilkInst.buffer).approve(usdsInst.usds, controllerInst.almProxy, type(uint256).max);
+    }
+
+    function makerInit(address psm, address almProxy) internal {
+        IPSMLike(psm).kiss(almProxy);  // To allow using no fee functionality
+    }
+
+}
+
+library ForeignControllerInit {
+
+    // Avoid stack too deep
+    struct AddressParams {
+        address freezer;
+        address relayer;
+        address usdc;
+        address usds;
+        address susds;
+    }
+
+    struct RateLimitData {
+        uint256 maxAmount;
+        uint256 slope;
+    }
+
+    struct InitRateLimitData {
+        RateLimitData usdcDepositData;
+        RateLimitData usdsDepositData;
+        RateLimitData susdsDepositData;
+        RateLimitData usdcWithdrawData;
+        RateLimitData usdsWithdrawData;
+        RateLimitData susdsWithdrawData;
+    }
+
+    function init(
+        AddressParams      memory params,
+        ControllerInstance memory controllerInst,
+        InitRateLimitData  memory data
+    )
+        internal
+    {
```

Comment:
Same here for sanity checks to make sure constructor arguments are set correctly without having to manually check.

Reply:
Added